addons/web/static/src/legacy/js/widgets/colorpicker.js

```diff
@@ -426,7 +426,7 @@ var ColorpickerWidget = Widget.extend({
      * @private
      * @param {Event} ev
      */
-    _onChangeInputs: function (ev) {
+    _onChangeInputs: _.throttle(function (ev) {
```

Comment:
If I get it correctly the issue here is that `_onChangeInputs` is called twice once from `_onKeypress`, another from `change` event?

Reply:
@deso-odoo Yes, you are correct. When the Enter key is pressed, it triggers the `_onKeypress` event, which subsequently calls `_onChangeInputs` to initiate the process of applying a color. However, during this process, `change` event, is also triggered, leading to another call to `_onChangeInputs`. Consequently, two processes are running concurrently to apply color to the current selection. To solve this issue, I have implemented `throttle` to ensure that the same process doesn't start again before the first one completes.

Reply:
The question is then: why is it a problem for the editor while it is not a problem for anyone else?

Reply:
@dmo-odoo Are you referring to a module where the editor is not used? If so, do you know of any such module that uses the `colorpicker`, or are you talking about the website and mass_mailing modules?

If you are referring to the website and mass_mailing modules, I can replicate the first issue in mass_mailing. On the website, instead of a traceback error, applying color to the last few words of a paragraph creates an extra `font` tag after the selected word, containing a zero-width space (ZWS).

Also, I am able to produce the second issue in both the mass_mailing and website.

Reply:
Ok I see what you mean. What I still don't understand though, is why it causes an issue in the editor. The editor should be able to receive two very fast commands about applying colors without crashing. Here you fix the fact that such concurrent commands are triggered in the first place, but shouldn't we fix the code of the editor so that if it happens again due to other reasons then we won't get a traceback?

Reply:
> what I still don't understand is why it causes an issue in the editor in the first place. The editor should be able to receive two commands about applying colors without crashing.

The problem arises because when [sel.addRange](https://github.com/odoo/odoo/blob/15.0/addons/web_editor/static/lib/odoo-editor/src/utils/utils.js#L612) is called within `setSelection`, it pauses the code execution and triggers the `change` event. Once the `change` event finishes, the code resumes from where it left off, which is `sel.extend(focusNode, focusOffset)`. However, by this point, the color has already been applied to the text from the change event, but `focusOffset` still refers to the old offset where the color hasn't been applied yet, leading to a traceback when attempting to extend the selection.

> To fix this in editor

I have tried adding await for [historyResetLatestComputedSelection](https://github.com/odoo/odoo/blob/15.0/addons/web_editor/static/src/js/wysiwyg/wysiwyg.js#L1613
) and it works.

Another approach I have tried is to directly create a range with `focusNode` and `focusOffset` instead of using `extend`. However, this approach creates issues with the selection direction.

Reply:
> > what I still don't understand is why it causes an issue in the editor in the first place. The editor should be able to receive two commands about applying colors without crashing.
> 
> The problem arises because when [sel.addRange](https://github.com/odoo/odoo/blob/15.0/addons/web_editor/static/lib/odoo-editor/src/utils/utils.js#L612) is called within `setSelection`, it pauses the code execution and triggers the `change` event. Once the `change` event finishes, the code resumes from where it left off, which is `sel.extend(focusNode, focusOffset)`.

Do you mean the `selectionchange` event, which would be weird since it's triggered asynchronously, or the `change` event of the colorpicker input, which would also be kind of weird since the content of the input does not change from `sel.addRange` ?

Reply:
I am referring to the `change` event of the colorpicker input. This happens because initially, before the `sel.addRange` is executed our selection resides within the input box. However, when `sel.addRange` is executed the selection moves to the text we want to color. This action causes us to lose focus on the input box, consequently triggering the change event.

Reply:
I see. And what about the `_computeHistorySelection` below ?

Reply:
In the `color_hover` event, the `processAndApplyColor` function applies color to the text but fails to trigger the `selectionChange` event. Because of this `_latestComputedSelection` does not get updated. Consequently, when the `color_picked` event triggers and executes `historyResetLatestComputedSelection`, it resets the selection incorrectly, as `_latestComputedSelection` wasn't updated in `color_hover`. As a result, our selection shifts slightly forward. Therefore, I have included the `_computeHistorySelection` here to ensure that `_latestComputedSelection` gets updated correctly.

Reply:
I see, thanks for the explanation! Since we removed the library that provided `_.throttle` in recent versions, would you mind using the `debounce` helper that lies in `timing.js` with the `immediate` parameter set to `true`? It's the only helper that does what you want to do in `master` and it is already there in 15.0 so I think it would minimize issues during fw-ports.

Reply:
@dmo-odoo I have implemented the requested changes.

Reply:
@dhba-odoo Thank you! I would like to set a limit for the forward port before merging this. Could you have a look and see in which version the bug does not appear anymore please?

Reply:
@dmo-odoo The traceback issue persists up to version 16.0. From version 17.0 onwards, the traceback does not occur, but a zero width space (zws) is added after the selected text when applying color to it.

Regarding the issue where color is applied to text that is not selected, this problem occurs across all versions.